### ISSUE TITLE
samtools fastq skips over reads with no bases/qualities

### ISSUE THREAD

**Issue by nh13** [samtools/samtools#1799]
#### Are you using the latest version of [samtools](https://github.com/samtools/samtools/releases/latest) and [HTSlib](https://github.com/samtools/htslib/releases/latest)? If not, please specify.

```
samtools 1.16.1
Using htslib 1.16
```

 #### Please describe your environment.

```
uname -sr
Darwin 21.6.0
```

#### Please specify the steps taken to generate the issue, the command you are running and the relevant output.

Input SAM
```
@HD	VN:1.6	SO:unsorted	GO:query
@PG	ID:samtools	PN:samtools	VN:1.14
query	77	*	0	0	*	*	0	0	GATTACA	IIIIIII
query	141	*	0	0	*	*	0	0	*	*
```

I am expecting two reads, but only get one:

```
samtools fastq -N tmp.sam
[M::bam2fq_mainloop] discarded 0 singletons
[M::bam2fq_mainloop] processed 2 reads
@query/1
GATTACA
+
IIIIIII
```

This is an issue when piping the output of `samtools fastq` into `bwa mem -p`, where `bwa mem` is expecting read pairs but treats this is a singleton.  I was expecting this output:

```
@query/1
GATTACA
+
IIIIIII
@query/2

+

```

**Comment by jkbonfield** [comment]
Hmm, I'm not convinced the expected output is helpful.  Would bwa accept an empty sequence?  If so what's it going to do with it?  I guess it'll just emit it as unmapped in the same manner as it was in the input SAM.

**Comment by jkbonfield** [comment]
I can confirm it was 227d082a1cc070dc97bb6278a38351f1216b85e1 that changed this (part of https://github.com/samtools/samtools/pull/1323).  That was a major overhaul of samtools fastq and fixed many bugs.

I can't yet remember whether removing empty records was a conscious decision or not.  I'll investigate.

**Comment by jkbonfield** [comment]
Found it:  https://github.com/samtools/htslib/blob/develop/sam.c#L4260 is the culprit.
```
    if (len == 0) return 0;
```
Comment out that and you get the previous behaviour.

This was my htslib change, where I added fastq as a native input and output format to replace the mess of adhoc fastq handling with something dealt with at the core.  I clearly decided zero length fastq was a problem best avoided, but I'm not sure now if it's the right thing to do.

As you're reporting this, you obvious have your own views.  Why should we permit zero length fastq?  What should tools do when reading it?  (For the record, `samtools import` does the right thing here and gives you back the original SAM file)
